src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -151,6 +155,59 @@ static BoundNode unwrapListElement(BoundCollectionExpression node, BoundNode ele
             }
         }
 
+        // If we have something like `List<int> l = [.. someEnumerable]`
+        // try rewrite it using `Enumerable.ToList` member if possible
+        private bool TryRewriteSingleElementSpreadToList(BoundCollectionExpression node, TypeWithAnnotations listElementType, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            result = null;
+
+            if (node.Elements is not [BoundCollectionExpressionSpreadElement singleSpread])
+            {
+                return false;
+            }
+
+            if (!ShouldUseAddRangeOrToListMethod(singleSpread.Expression.Type!, singleSpread.EnumeratorInfoOpt?.GetEnumeratorInfo.Method, listElementType))
+            {
+                return false;
+            }
+
+            if (TryGetWellKnownTypeMember(node.Syntax, WellKnownMember.System_Linq_Enumerable__ToList, out MethodSymbol? toListGeneric, isOptional: true))
+            {
+                var toListOfElementType = toListGeneric.Construct([listElementType]);
+                result = _factory.Call(receiver: null, toListOfElementType, singleSpread.Expression);
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool ShouldUseAddRangeOrToListMethod(TypeSymbol spreadType, MethodSymbol? getEnumeratorMethod, TypeWithAnnotations elementType)
+        {
+            var iEnumerableOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var iEnumerableOfElementType = iEnumerableOfTType.Construct([elementType]);
+
+            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+            var conversion = _compilation.Conversions.ClassifyImplicitConversionFromType(spreadType, iEnumerableOfElementType, ref discardedUseSiteInfo);
+            if (conversion.Kind is not (ConversionKind.Identity or ConversionKind.ImplicitReference))
+            {
+                return false;
+            }
+
+            if (getEnumeratorMethod?.ReturnType.IsValueType == true)
+            {
+                var iCollectionOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_ICollection_T);
+                var iCollectionOfElementType = iCollectionOfTType.Construct([elementType]);
+
+                if (!spreadType.ImplementsInterface(iCollectionOfElementType, ref discardedUseSiteInfo))
```

Comment:
>ImplementsInterface [](http://example.com/codeflow?start=32&length=19)

`ImplementsInterface` appears to be insufficient since it uses `TypeCompareKind.ConsiderEverything2`. (I overlooked this when reviewing [https://github.com/dotnet/roslyn/pull/74630](https://github.com/dotnet/roslyn/pull/74630).) For instance, it looks like we not use `List<object?>.AddRange(IEnumerable<object?>)` in the following where `MyCollection<T>` has a `struct` enumerator and implements `ICollection<T>`:
```
static List<object?> M(MyCollection<object> c) => [..c];
```

Reply:
I don't see `ImplementsInterface` overload that takes `TypeCompareKind` as an argument. Is there an existing helper or do we need to create a new one? Also what `TypeCompareKind` should we use here? My assumption is `CLRSignatureCompareOptions`, does it look correct?

Reply:
`CLRSignatureCompareOptions` is the right option to ignore differences that are not relevant at runtime. But I don't think we should be using `ImplementsInterface()` here. We should probably use `ConversionsBase.ClassifyBuiltInConversion(spreadType, iCollectionOfElementType)` or similar conversion method.

Reply:
It looks like `ImplementsInterface` affects the changes from [https://github.com/dotnet/roslyn/pull/74630](https://github.com/dotnet/roslyn/pull/74630) as well (see [https://github.com/dotnet/roslyn/pull/74769#issuecomment-2308489340](https://github.com/dotnet/roslyn/pull/74769#issuecomment-2308489340)). We should consider replacing the existing call to `ImplementsInterface` in a separate, dedicated PR.

Reply:
Logged https://github.com/dotnet/roslyn/issues/74894 to cover the existing issue.